### ISSUE TITLE
upgrade to Swig 4.2 gives assertion failure (replace_count), function Swig_smartptr_upcast, file Modules/utils.cxx, line 177

### ISSUE THREAD

**Issue by KrisThielemans** [swig/swig#2768]
I’m trying to update our conda-forge feedstock to the latest version of our STIR software. However, all MacOS builds fail when calling swig to generate the Python wrapper:
 ```
Assertion failed: (replace_count), function Swig_smartptr_upcast, file Modules/utils.cxx, line 177.
```
However, Linux and Windows builds work fine. All of them use conda-forge builds of swig 4.2.0. Downgrading swig to 4.1 for the MacOS build only makes this problem disappear. The relevant PR is https://github.com/conda-forge/stir-feedstock/pull/105 with a bit more detail, although nothing relevant I think.

@wsfulton wrote in response to my email to the list
> Grab the preprocessed output by adding -E:
>     swig4.0 -E -python -outdir . -c++ -builtin -interface _stir -o stirPYTHON_wrap.cxx stir.i > stir.preprocssed.i
> and create a Github issue. We should be able to recreate it with this. assert is only used with debug builds, so that should explain why it is different on different systems if SWIG is built differently.  I think just with or without -g

I'm not sure why the MacOS conda-forge build is a debug build and the others not, but I haven't checked the SWIG feedstock.

I also didn't build a SWIG debug build myself yet (on Linux). sorry.

For the preprocessing, I ran the following
```
swig -python  -DSTART_NAMESPACE_STIR="namespace stir {" -DEND_NAMESPACE_STIR="}" -DSTIR_DEPRECATED="" -DSTIR_TOF=1 -DSTIR_VERSION=060000 -DHAVE_JSON -DHAVE_ITK -DHAVE_parallelproj -c++ -builtin -DDOXY2SWIG_XML_INCLUDE_FILE="\"/home/sirfuser/devel/build/STIR/RelWithDebug-gcc8/src/swig/STIR_DOXY2SWIG.i\"" -interface _stir -I/home/sirfuser/devel/STIR2/src/include -E  /home/sirfuser/devel/STIR2/src/swig/stir.i
```

[Preprocessed output file](https://github.com/swig/swig/files/13997335/stir.swig4.2.0.txt)

Somewhat similar issue is #2755 but it looks like it is a different assertion.

**Comment by ojwb** [comment]
I notice William said to use `swig4.0 -E -python -outdir . -c++ -builtin -interface _stir -o stirPYTHON_wrap.cxx stir.i > stir.preprocssed.i` but your output seems to be from SWIG 4.2.0:

```
%includefile "/home/sirfuser/mambaforge/envs/swig4/share/swig/4.2.0/swig.swg" %beginfile
```

> I also didn't build a SWIG debug build myself yet (on Linux). sorry.

A debug build is the default (assertions are on unless you define `NDEBUG` when building SWIG).

Backtrace from gdb (#5 on is the relevant part):

```
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007ffff7aa815f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007ffff7a5a472 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff7a444b2 in __GI_abort () at ./stdlib/abort.c:79
#4  0x00007ffff7a443d5 in __assert_fail_base (fmt=0x7ffff7bb8dc8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x5555556adc4f "replace_count", 
    file=file@entry=0x5555556adc3d "Modules/utils.cxx", line=line@entry=177, 
    function=function@entry=0x5555556c8458 "SwigType* Swig_smartptr_upcast(SwigType*, SwigType*, SwigType*)") at ./assert/assert.c:92
#5  0x00007ffff7a533a2 in __assert_fail (assertion=0x5555556adc4f "replace_count", file=0x5555556adc3d "Modules/utils.cxx", line=177, 
    function=0x5555556c8458 "SwigType* Swig_smartptr_upcast(SwigType*, SwigType*, SwigType*)") at ./assert/assert.c:101
#6  0x0000555555672947 in Swig_smartptr_upcast (smart=smart@entry=0x7ffff2c2ee70, classname=classname@entry=0x7ffff2c2da70, 
    baseclassname=baseclassname@entry=0x7ffff2c2c110) at Modules/utils.cxx:177
#7  0x0000555555670fb7 in TypePass::cplus_inherit_types_impl (this=this@entry=0x7fffffffdce0, first=first@entry=0x7ffff19cb210, cls=0x7ffff19cb210, cls@entry=0x0, 
    clsname=clsname@entry=0x7ffff2c2da70, bases=bases@entry=0x5555556a5882 "bases", baselist=baselist@entry=0x5555556a7090 "baselist", ispublic=1, cast=0x0)
    at Modules/typepass.cxx:269
#8  0x0000555555671942 in TypePass::cplus_inherit_types (this=0x7fffffffdce0, first=0x7ffff19cb210, cls=0x0, clsname=0x7ffff2c2da70, cast=0x0) at Modules/typepass.cxx:331
#9  0x0000555555671ea6 in TypePass::classDeclaration (this=0x7fffffffdce0, n=0x7ffff19cb210) at Modules/typepass.cxx:509
#10 0x00005555556143a3 in Dispatcher::emit_one (this=0x7fffffffdce0, n=0x7ffff19cb210) at Modules/lang.cxx:133
#11 0x00005555556147f1 in Dispatcher::emit_children (this=0x7fffffffdce0, n=<optimized out>) at Modules/lang.cxx:217
#12 0x000055555566f659 in TypePass::namespaceDeclaration (this=0x7fffffffdce0, n=0x7ffff19ca910) at Modules/typepass.cxx:635
#13 0x0000555555614493 in Dispatcher::emit_one (this=0x7fffffffdce0, n=0x7ffff19ca910) at Modules/lang.cxx:145
#14 0x00005555556147f1 in Dispatcher::emit_children (this=0x7fffffffdce0, n=<optimized out>) at Modules/lang.cxx:217
#15 0x000055555561460b in Dispatcher::emit_one (this=0x7fffffffdce0, n=0x7ffff151d570) at Modules/lang.cxx:171
#16 0x00005555556147f1 in Dispatcher::emit_children (this=0x7fffffffdce0, n=<optimized out>) at Modules/lang.cxx:217
#17 0x000055555561460b in Dispatcher::emit_one (this=0x7fffffffdce0, n=0x7fffefbbceb0) at Modules/lang.cxx:171
#18 0x00005555556147f1 in Dispatcher::emit_children (this=0x7fffffffdce0, n=<optimized out>) at Modules/lang.cxx:217
#19 0x000055555561460b in Dispatcher::emit_one (this=0x7fffffffdce0, n=0x7fffefa949b0) at Modules/lang.cxx:171
#20 0x00005555556147f1 in Dispatcher::emit_children (this=0x7fffffffdce0, n=<optimized out>) at Modules/lang.cxx:217
#21 0x00005555556702c5 in TypePass::top (this=this@entry=0x7fffffffdce0, n=n@entry=0x7fffef95c7f0) at Modules/typepass.cxx:360
#22 0x000055555566f3b2 in TypePass::pass (n=0x7fffef95c7f0) at Modules/typepass.cxx:1106
#23 Swig_process_types (n=n@entry=0x7fffef95c7f0) at Modules/typepass.cxx:1113
#24 0x0000555555623adc in SWIG_main (argc=<optimized out>, argv=<optimized out>, tlm=<optimized out>) at Modules/main.cxx:1196
#25 0x00007ffff7a456ca in __libc_start_call_main (main=main@entry=0x55555556ea80 <main(int, char**)>, argc=argc@entry=4, argv=argv@entry=0x7fffffffdf18)
    at ../sysdeps/nptl/libc_start_call_main.h:58
#26 0x00007ffff7a45785 in __libc_start_main_impl (main=0x55555556ea80 <main(int, char**)>, argc=4, argv=0x7fffffffdf18, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffdf08) at ../csu/libc-start.c:360
#27 0x000055555556f101 in _start ()
```

**Comment by KrisThielemans** [comment]
> I notice William said to use `swig4.0 -E -python -outdir . -c++ -builtin -interface _stir -o stirPYTHON_wrap.cxx stir.i > stir.preprocssed.i` but your output seems to be from SWIG 4.2.0

`swig4.0` is actually 4.2.0 (as I noted in the STIR conda-forge PR). I can run with older versions of swig as well if that'd help.

> A debug build is the default (assertions are on unless you define NDEBUG when building SWIG).

I'm using the conda-forge build. On Linux and windows, it doesn't assert.

Unfortunately, the stack trace doesn't tell me anything. Hopefully it does for one of you!

**Comment by ojwb** [comment]
> `swig4.0` is actually 4.2.0 (as I noted in the STIR conda-forge PR). I can run with older versions of swig as well if that'd help.

I don't think so, just seemed like you'd maybe run the wrong version but actually it's confusing package name.

> Unfortunately, the stack trace doesn't tell me anything. Hopefully it does for one of you!

The failing assertion (and the whole function it is in) were added by @wsfulton in 2acdfd77e9e498d604e1dc0636cf675f9ce10244 so I think this is one for him.

**Comment by wsfulton** [comment]
I'll take a look at this next.

**Comment by wsfulton** [comment]
Problem area is /home/sirfuser/devel/STIR2/src/swig/stir_dataprocessors.i:50. Can you please provide the exact git hash in the stir code base and the github link to it that you used for the preprocessed output.

I think it is line 188800 of the preprocessed file:
```
%template(RPChainedDataProcessor3DFloat) stir::RegisteredParsingObject<
             stir::ChainedDataProcessor<stir::DiscretisedDensity<3,float> >,
             stir::DataProcessor<DiscretisedDensity<3,float> >,
             stir::DataProcessor<DiscretisedDensity<3,float> > >;
```
and something to do with a `%shared_ptr`
Might be something to do with DiscretisedDensity not being fully qualified as the assertion fails because SWIG is trying to make a substitution and fails because the type in the %template above is not fully qualified. I think there might be a `using namespace stir` at this point, I see comments like this:
```
   // TODO need this (bug in swig)
   // this bug occurs (only?) when using "%template(name) someclass;" inside the namespace
   // as opposed to "%template(name) stir::someclass" outside the namespace
   using namespace stir;
```
That might be out of date now, but a cut-down replication of that in a new issue would be useful too.

**Comment by wsfulton** [comment]
I've added in some calls to `normalize_type` for the smartptr feature handling to deal with `using namespace stir` and the unqualified `DiscretisedDensity` in the instantiation:
```swig
%template(RPChainedDataProcessor3DFloat) stir::RegisteredParsingObject<
             stir::ChainedDataProcessor<stir::DiscretisedDensity<3,float> >,
             stir::DataProcessor<DiscretisedDensity<3,float> >,
             stir::DataProcessor<DiscretisedDensity<3,float> > >;
```
yuck. 

Needs some more testing and possible tweaking before I commit it.

**Comment by KrisThielemans** [comment]
Thanks @wsfulton !

> Can you please provide the exact git hash in the stir code base and the github link to it that you used for the preprocessed output.

[434837c8e97a80b6402d6684bceaa849f44b4d8a](https://github.com/UCL/STIR/commit/434837c8e97a80b6402d6684bceaa849f44b4d8a)

> That might be out of date now, but a cut-down replication of that in a new issue would be useful too.

A lot of the code in the  STIR *.i  (and indeed the C++ code base) is quite old (20 years or so!). It therefore contains work-arounds for SWIG bugs that might no longer exist. One day, I'll get round to removing the work-arounds (probably when they stop working!). And of course, there will be inefficiencies here because of my lack of understanding of SWIG (for isntance, I'm fairly sure I over-use `%shared_ptr`.

> ```c++
> %template(RPChainedDataProcessor3DFloat) stir::RegisteredParsingObject<
>              stir::ChainedDataProcessor<stir::DiscretisedDensity<3,float> >,
>              stir::DataProcessor<DiscretisedDensity<3,float> >,
>              stir::DataProcessor<DiscretisedDensity<3,float> > >;
> ```
> yuck.

I'm assuming the "yuck" is that this is rather ugly and verbose, and I agree!

For the above lines, there are a few reasons that it's as ugly as it is.
- the last argument [is a default template argument](https://github.com/UCL/STIR/blob/59c3896524cd7669c0a81c5e5eea49368dff4f1e/src/include/stir/RegisteredParsingObject.h#L77-L78), but SWIG didn't handle that, so I had to repeat it. According to "6.18.3 Default template arguments" this should no longer be necessary.
- The `RPChainedDataProcessor3DFloat`   name should have been unnecessary, as this is still an abstract class. However, I do need its members. I guess this is a case of "not wrapping" but needing "instantiation", but using `%template()` didn't work.

Possibly these problems only exist(ed) because this class is case of [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern).

I realise I should now go ahead and test all this with recent SWIG and create some issues if the problems persist, but sadly time is in very short supply. I'll try to find some.

Anyway, this is all going beyond the original problem report, so I'm not sure where the best place is to discuss the obsolete work-arounds of the current STIR wrappers and potential SWIG bugs is, if you had the time for it.

**Comment by KrisThielemans** [comment]
@wsfulton add `stir::` to all `DiscretisedDensity` instances seems to have resolved this problem for me. See https://github.com/UCL/STIR/pull/1347 It went through my local Debug build of swig.Thanks!

**Comment by wsfulton** [comment]
> @wsfulton add `stir::` to all `DiscretisedDensity` instances seems to have resolved this problem for me.

Yes, fully qualifying template types is generally the recommendation in order to avoid some quirks and corner case bugs in the type system. In this case it should not have been necessary and I have fixed the problem in SWIG when handling 'feature:smart_ptr` used by `%shared_ptr`.

I had to really think about the scope in which the type within the 'feature:smartptr' needed to be resolved in as mentioned in the commit message.  The code that SWIG parses did not have the `using namespace stir` that was only seen by the C++ compiler. In the end, I decided the `using namespace stir` is not necessary because the template parameter `DiscretisedDensity` is resolved in the scope of 

```
stir::RegisteredParsingObject<
                              stir::ChainedDataProcessor<stir::DiscretisedDensity<3,float> >,
                              stir::DataProcessor<DiscretisedDensity<3,float> >,
                              stir::DataProcessor<DiscretisedDensity<3,float> > >;
```
which is within the `stir` namespace, so will resolve to `stir::DiscretisedDensity`.

It took quite a bit of effort to replicate the issue getting the ordering right, but https://github.com/swig/swig/blob/3be670e8db5b45195b670c91eac3fb413cf8427b/Examples/test-suite/cpp11_shared_ptr_crtp_upcast.i replicated the assert and is now a testcase.

>    The RPChainedDataProcessor3DFloat name should have been unnecessary, as this is still an abstract class. However, I do need its members. I guess this is a case of "not wrapping" but needing "instantiation", but using %template() didn't work.

%template() ought to have worked. If it still doesn't, that ought to go into a new issue.
